src/AbstractUri.php

```diff
@@ -362,33 +356,76 @@ protected static function urlEncodeMatch(array $matches): string
      *
      * @return string|null
      */
-    protected function formatHost($host)
+    protected function formatHost(string $host = null)
     {
-        static $pattern = '/^(?<name>[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)(?:\.(?&name))*$/i';
-
-        if ('' == $host || false !== strpos($host, ']')) {
+        if (null === $host || '' === $host) {
             return $host;
         }
 
+        static $pattern = '/^(?<name>[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)(?:\.(?&name))*$/i';
         if (!isset($host[253]) && preg_match($pattern, $host)) {
             return strtolower($host);
         }
 
-        $component = '';
-        $valid_ascii_label_characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-';
-        foreach (explode('.', mb_strtolower($host, 'UTF-8')) as $label) {
-            if (false !== strpos($label, '%')) {
-                $label = rawurldecode($label);
-            }
+        if (!is_host($host)) {
+            throw UriException::createFromInvalidHost($host);
+        }
 
-            if (strlen($label) !== strspn($label, $valid_ascii_label_characters)) {
-                $label = (string) idn_to_ascii($label, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46);
-            }
+        if (false !== strpos($host, ']')) {
```

Comment:
Any reason why this isn't `($host[0] ?? '') === '['`? Saves the function call and iterating the whole string internally.

Reply:
none I'll use this check instead